src/ir/possible-contents.h

```diff
@@ -0,0 +1,577 @@
+  // contents here will then include whatever content was possible in |other|.
+  void combine(const PossibleContents& other) {
+    // First handle the trivial cases of them being equal, or one of them is
+    // None or Many.
+    if (*this == other) {
+      return;
+    }
+    if (other.isNone()) {
+      return;
+    }
+    if (isNone()) {
+      value = other.value;
+      return;
+    }
+    if (isMany()) {
+      return;
+    }
+    if (other.isMany()) {
+      value = Many();
+      return;
+    }
+
+    auto type = getType();
+    auto otherType = other.getType();
+
+    if (!type.isRef() || !otherType.isRef()) {
+      // At least one is not a reference. We could in principle try to find
+      // ExactType here, say as the combination of two different constants, but
+      // as subtyping does not exist, Many is good enough anyhow, so do that.
+      value = Many();
+      return;
+    }
+
+    // Special handling for references from here.
+
+    // Nulls are always equal to each other, even if their types differ.
+    if (isNull() || other.isNull()) {
+      // If only one is a null, but the other's type is known exactly, then the
+      // combination is to add nullability (if the type is *not* known exactly,
+      // like for a global, then we cannot do anything useful here).
+      if (!isNull() && isTypeExact()) {
+        value = ExactType(Type(type.getHeapType(), Nullable));
+        return;
+      } else if (!other.isNull() && other.isTypeExact()) {
+        value = ExactType(Type(otherType.getHeapType(), Nullable));
+        return;
+      } else if (isNull() && other.isNull()) {
+        // Both are null. The result is a null, of the LUB.
+        auto lub = HeapType::getLeastUpperBound(type.getHeapType(),
+                                                otherType.getHeapType());
+        value = Literal::makeNull(lub);
+        return;
+      }
+    }
+
+    if (isTypeExact() && other.isTypeExact() &&
+        type.getHeapType() == otherType.getHeapType()) {
+      // We know the types here exactly, and even the heap types match, but
+      // there is some other difference that prevents them from being 100%
+      // identical (for example, one might be an ExactType and the other a
+      // Literal; or both might be ExactTypes and only one might be nullable).
+      // In these cases we can emit a proper ExactType here, adding nullability
+      // if we need to.
+      value = ExactType(Type(
+        type.getHeapType(),
+        type.isNullable() || otherType.isNullable() ? Nullable : NonNullable));
+      return;
+    }
+
+    // Nothing else possible combines in an interesting way; emit a Many.
+    value = Many();
+  }
+
+  bool isNone() const { return std::get_if<None>(&value); }
+  bool isLiteral() const { return std::get_if<Literal>(&value); }
+  bool isGlobal() const { return std::get_if<GlobalInfo>(&value); }
+  bool isExactType() const { return std::get_if<Type>(&value); }
```

Comment:
This would make these methods strictly more useful, since they could be used in all the same contexts as the `is*` variants, but could also be used in conditions that perform assignment such as in `getType` below.

```suggestion
  Literal* getLiteral() const { return std::get_if<Literal>(&value); }
  GlobalInfo* getGlobal() const { return std::get_if<GlobalInfo>(&value); }
  Type* getExactType() const { return std::get_if<Type>(&value); }
```

Reply:
Interesting, but I think it's slightly safer as it is, where `getLiteral` etc. will assert instead of returning null.

Reply:
Why? Any mistakes would turn into segfaults (at least on native builds), which are similar to assertions in terms of usefulness.

Reply:
I guess I don't feel strongly. One issue, though, is that `get_if` returns a pointer to something whose lifetime might end before the pointer is discarded. It's just not a safe/nice API, I guess is how I feel about it. To make it safe we could make copies of the things before returning them here (they all fit in register or two), but at that point I feel like it's making the code less clear...

Reply:
Yes, another alternative is to return a copy in a `std::optional`. That seems like overkill though. By far the most common use case I've seen for this kind of code is

```
if (auto* val = var.getSomeVariant()) {
  // Use val without it escaping
}
```

Reply:
That's definitely the most common case. I'm worried about bugs in the less common paths though. And it is a simpler mental model to not have to think about the lifetime of the pointer returned here, by not returning a pointer.

If you feel strongly about this, though, then I think a copy + std::optional might be a reasonable compromise? (I should benchmark it if we want to go that way, though.)

Reply:
Oh, a downside of `std::optional`, however, is the lack of asserts... I've been bitten by this a few times already. Doing `*thing` on an empty optional is undefined behavior, and in practice very strange things happen :cry:

Reply:
Too bad we're not using Rust, where neither lifetimes nor accidental dereferences can ever be problems :crab: :cry: :crab: